### PR TITLE
refactor(app): remove unused amd requires

### LINKED ISSUE #2392: Investigate unused amd requires
Per @shane-tomlinson's https://github.com/mozilla/fxa-content-server/pull/2377#discussion_r29797070

> _"I don't think `p` is used anywhere in this module."_

---

I was playing w/ [**grunt-amdcheck**](https://www.npmjs.com/package/grunt-amdcheck) and noticed some interesting results ~~(but lots of noise)~~ (fixed in grunt-amdcheck@1.3.0).

Here is my _grunttasks/**amdcheck.js**_ file:

```js
module.exports = function (grunt) {
  'use strict';

  grunt.config('amdcheck', {
    app: {
      options: {
        removeUnusedDependencies: false
      },
      files: [{
        expand: true,
        cwd: 'app/',
        src: [
          '**/*.js',
          '!bower_components/**'
        ]
      }]
    }
  });
}
```

And here is the ~~difficult to read~~ output:

```
$ grunt amdcheck

Running "amdcheck:app" (amdcheck) task

app/scripts/lib/channels/fx-desktop.js (1 module)
Unused paths: lib/auth-errors

app/scripts/lib/config-loader.js (1 module)
Unused paths: underscore

app/scripts/lib/fxa-client.js (1 module)
Unused paths: underscore, lib/xhr

app/scripts/lib/oauth-client.js (1 module)
Unused paths: lib/promise, lib/session

app/scripts/models/account.js (1 module)
Unused paths: lib/auth-errors

app/scripts/models/auth_brokers/iframe.js (1 module)
Unused paths: jquery, lib/promise, lib/channels/iframe, lib/auth-errors

app/scripts/models/auth_brokers/mixins/channel.js (1 module)
Unused paths: underscore

app/scripts/models/reliers/oauth.js (1 module)
Unused paths: lib/promise

app/scripts/models/user.js (1 module)
Unused paths: lib/assertion, lib/oauth-client

app/scripts/views/complete_reset_password.js (1 module)
Unused paths: lib/validate

app/scripts/views/complete_sign_up.js (1 module)
Unused paths: lib/validate, lib/promise

app/scripts/views/confirm.js (1 module)
Unused paths: underscore

app/scripts/views/confirm_reset_password.js (1 module)
Unused paths: jquery, lib/constants

app/scripts/views/marketing_snippet_ios.js (1 module)
Unused paths: lib/constants

app/scripts/views/ready.js (1 module)
Unused paths: lib/session, lib/xss, lib/strings, lib/auth-errors, lib/promise

app/scripts/views/settings/avatar.js (1 module)
Unused paths: underscore

app/scripts/views/settings/avatar_camera.js (1 module)
Unused paths: canvasToBlob

app/scripts/views/settings/avatar_change.js (1 module)
Unused paths: underscore, lib/url

app/scripts/views/settings/avatar_crop.js (1 module)
Unused paths: underscore

app/scripts/views/sign_in.js (1 module)
Unused paths: lib/validate

app/scripts/views/sign_up.js (1 module)
Unused paths: lib/session, lib/strings

app/tests/spec/lib/app-start.js (1 module)
Unused paths: models/reliers/base

app/tests/spec/lib/assertion.js (1 module)
Unused paths: sinon, lib/constants, vendor/jwcrypto/lib/algs/rs

app/tests/spec/lib/channels/duplex.js (1 module)
Unused paths: lib/auth-errors

app/tests/spec/lib/channels/fx-desktop.js (1 module)
Unused paths: lib/auth-errors

app/tests/spec/lib/channels/iframe.js (1 module)
Unused paths: lib/auth-errors

app/tests/spec/lib/channels/null.js (1 module)
Unused paths: chai

app/tests/spec/lib/channels/web.js (1 module)
Unused paths: router, views/sign_in

app/tests/spec/lib/cropper.js (1 module)
Unused paths: jquery

app/tests/spec/lib/fxa-client.js (1 module)
Unused paths: lib/session

app/tests/spec/lib/mailcheck.js (1 module)
Unused paths: lib/promise

app/tests/spec/lib/oauth-client.js (1 module)
Unused paths: jquery, ../../lib/helpers

app/tests/spec/lib/profile-client.js (1 module)
Unused paths: jquery

app/tests/spec/lib/router.js (1 module)
Unused paths: underscore

app/tests/spec/lib/url.js (1 module)
Unused paths: underscore

app/tests/spec/lib/validate.js (1 module)
Unused paths: underscore

app/tests/spec/models/auth_brokers/base.js (1 module)
Unused paths: sinon

app/tests/spec/models/auth_brokers/fx-desktop.js (1 module)
Unused paths: underscore, lib/constants

app/tests/spec/models/auth_brokers/iframe.js (1 module)
Unused paths: lib/auth-errors

app/tests/spec/models/auth_brokers/web-channel.js (1 module)
Unused paths: lib/auth-errors

app/tests/spec/models/notifications.js (1 module)
Unused paths: lib/promise

app/tests/spec/models/profile-image.js (1 module)
Unused paths: sinon, lib/promise

app/tests/spec/models/verification/base.js (1 module)
Unused paths: sinon

app/tests/spec/models/verification/reset-password.js (1 module)
Unused paths: sinon

app/tests/spec/models/verification/sign-up.js (1 module)
Unused paths: sinon

app/tests/spec/views/change_password.js (1 module)
Unused paths: underscore

app/tests/spec/views/confirm_reset_password.js (1 module)
Unused paths: lib/session

app/tests/spec/views/cookies_disabled.js (1 module)
Unused paths: lib/promise

app/tests/spec/views/coppa/coppa-date-picker.js (1 module)
Unused paths: underscore, lib/auth-errors, lib/promise

app/tests/spec/views/force_auth.js (1 module)
Unused paths: lib/session

app/tests/spec/views/mixins/avatar-mixin.js (1 module)
Unused paths: backbone

app/tests/spec/views/mixins/back-mixin.js (1 module)
Unused paths: underscore

app/tests/spec/views/mixins/password-mixin.js (1 module)
Unused paths: backbone

app/tests/spec/views/mixins/timer-mixin.js (1 module)
Unused paths: backbone

app/tests/spec/views/oauth_sign_up.js (1 module)
Unused paths: lib/auth-errors

app/tests/spec/views/ready.js (1 module)
Unused paths: lib/promise

app/tests/spec/views/reset_password.js (1 module)
Unused paths: lib/session

app/tests/spec/views/settings.js (1 module)
Unused paths: underscore, lib/constants

app/tests/spec/views/settings/avatar.js (1 module)
Unused paths: underscore

app/tests/spec/views/settings/avatar_camera.js (1 module)
Unused paths: underscore

app/tests/spec/views/settings/avatar_change.js (1 module)
Unused paths: underscore

app/tests/spec/views/settings/avatar_crop.js (1 module)
Unused paths: underscore, draggable, lib/constants

app/tests/spec/views/settings/avatar_gravatar.js (1 module)
Unused paths: underscore

app/tests/spec/views/sign_up.js (1 module)
Unused paths: underscore, moment


Total unused dependencies: 89 in 64 files.
Total processed files: 251

Done, without errors.


Execution Time (2015-05-12 17:23:13 UTC)
loading tasks  3.7s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 56%
amdcheck:app   2.9s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 44%
Total 6.6s
```

### PR DESCRIPTION
**Pull request by riadhchtara**
Fixes #2392

### CODE REVIEW

**Comment by vladikoff** [comment]
Thanks for the pull request!

This needs to be rebase against the `master` branch.

File: grunttasks/lint.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ module.exports = function (grunt) {
   'use strict';
 
   grunt.registerTask('lint', 'lint all the things', [
+    'amdcheck',
```
Comment (pdehaan) [inline grunttasks/lint.js]: I think this should probably be moved to [/grunttasks/concurrent.js:23-27](https://github.com/mozilla/fxa-content-server/blob/287179aaeffe1cdc7640745a93e54eedf2d975d3/grunttasks/concurrent.js#L23-L27).

File: grunttasks/amdcheck.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+module.exports = function (grunt) {
+  'use strict';
+
+  grunt.config('amdcheck', {
+    app: {
+      options: {
+        removeUnusedDependencies: false,
+        strict:true
```
Comment (pdehaan) [inline grunttasks/amdcheck.js]: nit: add a space between the ":" and "true" -- `strict: true`

File: grunttasks/amdcheck.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      },
+      files: [{
+        expand: true,
+        cwd: 'app/',
+        src: [
+          '**/*.js',
+          '!bower_components/**'
```
Comment (pdehaan) [inline grunttasks/amdcheck.js]: I admit, I was guessing here.
We may want to exclude more paths.
For example, `!scripts/vendor/**`

/cc @vladikoff @shane-tomlinson

Comment (vladikoff) [inline grunttasks/amdcheck.js]: can we include, `scripts` and `tests`, and exclude `scripts/vendor` ?

Comment (vladikoff) [inline grunttasks/amdcheck.js]: This way we do not ask ALL the folders in `app`.

Comment (pdehaan) [inline grunttasks/amdcheck.js]: Ha! You really do hate my lazy globs.

@vladikoff is of course 1000% correct. You can probably simply it to something like this:
```js
src: [
  '{scripts,tests}/**/*.js',
  '!scripts/vendor/**'
]
```

File: grunttasks/lint.js (hunk 1 above)
Comment (riadhchtara) [inline grunttasks/lint.js]: done!

File: grunttasks/amdcheck.js (hunk 2 above)
Comment (riadhchtara) [inline grunttasks/amdcheck.js]: done!

File: grunttasks/amdcheck.js (hunk 3 above)
Comment (riadhchtara) [inline grunttasks/amdcheck.js]: done!

**Comment by riadhchtara** [comment]
hi guys, thanks for the review
I fixed the issues and done a rebase.
Could you please review it again?

**Comment by vladikoff** [comment]
at first glance, some warnings:
```
       Warning: Running "amdcheck:app" (amdcheck) task
    app/scripts/head/startup-styles.js (1 module)
    Unused paths: lib/environment
    
    app/scripts/lib/mailcheck.js (1 module)
    Unused paths: mailcheck
    
    app/tests/spec/lib/assertion.js (1 module)
    Unused paths: vendor/jwcrypto/lib/algs/rs
    
    app/tests/spec/views/settings/avatar_crop.js (1 module)
    Unused paths: draggable
    
    
    Total unused paths: 4 in 4 files.
    Total unused dependencies: 1 in 1 files.
    Total processed files: 239
    Warning: Task "amdcheck:app" failed. Use --force to continue.
    
    Aborted due to warnings.
```

**Comment by riadhchtara** [comment]
Unfortunately those modules cannot be removed from those files, I think because that there used in other places.
When I remove only mailcheck, the follwing tests fails
lib/mailcheck
    works with attached elements and changes values ‣
     element.mailcheck is not a function
   works with attached elements and can be dismissed ‣
     element.mailcheck is not a function

views/sign_up
    suggestEmail
        suggests emails via a tooltip ‣
           element.mailcheck is not a function

**Comment by riadhchtara** [comment]
@vladikoff : could you please have a look to my previous comment?
Thanks

**Comment by vladikoff** [comment]
@riadhchtara for those exceptions can we try using the `/* exceptsPaths: [dependency_here] */ feature.

See https://github.com/mehdishojaei/grunt-amdcheck#exceptspaths
This is similar to what we did with `maxcomplexity`

**Comment by riadhchtara** [comment]
Hi @vladikoff : thanks for the suggestion. everything is working now.
Cold you please review the patch again?

**Comment by riadhchtara** [comment]
@vladikoff damn! I have just rebased, and there is a new conflict
There are so many files changed in the PR so the probability of conflict is huge.

**Comment by riadhchtara** [comment]
it would be awesome

**Pull request closed without merging**